File: backend/build_meta.py
from __future__ import annotations

import os
from collections.abc import Mapping
from datetime import datetime, timezone


APP_VERSION = "0.9.1"
PROTOCOL_VERSION = "1.0"
SHA_ENVIRONMENT_KEYS = (
    "VITE_GIT_SHA",
    "VERCEL_GIT_COMMIT_SHA",
    "RAILWAY_GIT_COMMIT_SHA",
    "GITHUB_SHA",
    "SOURCE_VERSION",
)
ENVIRONMENT_KEYS = (
    "VERCEL_ENV",
    "RAILWAY_ENVIRONMENT_NAME",
    "RAILWAY_ENVIRONMENT",
    "APP_ENV",
)


def build_metadata(environment: Mapping[str, str] | None = None) -> dict[str, str]:
    environment = os.environ if environment is None else environment
    git_sha = next((candidate for key in SHA_ENVIRONMENT_KEYS if (candidate := environment.get(key, "").strip()) and is_full_commit_sha(candidate)), "dev")
    build_environment = next(
        (environment[key].strip() for key in ENVIRONMENT_KEYS if environment.get(key, "").strip()),
        "development",
    )
    return {
        "schema_version": "build-meta.v1",
        "app_version": APP_VERSION,
        "protocol_version": PROTOCOL_VERSION,
        "git_sha": git_sha,
        "build_environment": build_environment,
        "build_time": build_time(environment),
    }


def is_full_commit_sha(value: str) -> bool:
    return len(value) in (40, 64) and all(character in "0123456789abcdefABCDEF" for character in value)


def build_time(environment: Mapping[str, str]) -> str:
    source_date_epoch = environment.get("SOURCE_DATE_EPOCH", "").strip()
    if source_date_epoch.isdigit():
        return (
            datetime.fromtimestamp(int(source_date_epoch), tz=timezone.utc)
            .isoformat(timespec="milliseconds")
            .replace("+00:00", "Z")
        )
    return environment.get("BUILD_TIME", "").strip() or "unknown"

File: frontend/scripts/build-meta.mjs
import { execFileSync } from 'node:child_process'

const SHA_ENVIRONMENT_KEYS = [
  'VITE_GIT_SHA',
  'VERCEL_GIT_COMMIT_SHA',
  'RAILWAY_GIT_COMMIT_SHA',
  'GITHUB_SHA',
  'SOURCE_VERSION',
]

export function resolveGitSha(environment = process.env, resolveHead = gitHead) {
  for (const key of SHA_ENVIRONMENT_KEYS) {
    const value = environment[key]?.trim()
    if (isFullCommitSha(value)) return value
  }
  const head = resolveHead()
  return isFullCommitSha(head) ? head : 'dev'
}

export function isFullCommitSha(value) {
  return typeof value === 'string' && /^(?:[0-9a-f]{40}|[0-9a-f]{64})$/i.test(value)
}

export function buildEnvironment(environment = process.env) {
  return (
    environment.VITE_BUILD_ENV?.trim() ||
    environment.VERCEL_ENV?.trim() ||
    environment.RAILWAY_ENVIRONMENT_NAME?.trim() ||
    environment.RAILWAY_ENVIRONMENT?.trim() ||
    environment.NODE_ENV?.trim() ||
    'development'
  )
}

export function buildTime(environment = process.env, now = new Date()) {
  const sourceDateEpoch = environment.SOURCE_DATE_EPOCH?.trim()
  if (sourceDateEpoch && /^\d+$/.test(sourceDateEpoch)) {
    return new Date(Number(sourceDateEpoch) * 1000).toISOString()
  }
  return environment.BUILD_TIME?.trim() || now.toISOString()
}

export function buildMeta({ appVersion, environment = process.env, now = new Date(), resolveHead = gitHead }) {
  return {
    schema_version: 'build-meta.v1',
    app_version: appVersion,
    protocol_version: '1.0',
    git_sha: resolveGitSha(environment, resolveHead),
    build_environment: buildEnvironment(environment),
    build_time: buildTime(environment, now),
  }
}

function gitHead() {
  try {
    return execFileSync('git', ['rev-parse', 'HEAD'], { encoding: 'utf8' }).trim() || null
  } catch {
    return null
  }
}
